onnxruntime_extensions/tools/pre_post_processing/steps/nlp.py

```diff
@@ -0,0 +1,508 @@
+            we support taking the other parameter as optional and use the default value such as the same usage in Hugging-Face.
+
+        Args:
+            tokenizer_map: some essential infos, If you export tokenizer from hugging-face,
+            such as "tokenizer = transformers.AutoTokenizer.from_pretrained(model_name)"
+            You may need to provide the following:
+                tokenizer_param = TokenizerParam(vocab_size=tokenizer.vocab_size,
+                                    bos_token_id=tokenizer.bos_token_id,
+                                    eos_token_id = tokenizer.eos_token_id,)
+            name: Optional name of step. Defaults to 'SentencePieceTokenizer'
+
+        """
+        super().__init__(
+            ["inputs", "nbest_size", "alpha", "add_bos", "add_eos", "reverse"], ["input_ids", "attention_mask"], name
+        )
+        self._domain = "com.microsoft.extensions"
+        self._tokenizer_param = tokenizer_param
+
+    def _create_graph_for_step(self, graph: onnx.GraphProto, onnx_opset: int):
+        """
+        the first input is required, the other four are optional. we don't expect user to provide all of them.
+        When created inputs, every input has a name which corresponds to the name in the graph's input.
+        So, If you want to take the default value, just don't provide it.
+        """
+        graph_input_names = [inp.name for inp in graph.output]
+        assert self.input_names[0] in graph_input_names
+
+        # input text
+        input_type_str0, input_shape_str0 = self._get_input_type_and_shape_strs(graph, 0)
+        """
+        SentencePieceTokenizer has actually 5 inputs in definition, so this complexity checks are make the other four args 
+        "nbest_size", "alpha", "add_bos", "add_eos", "reverse" as optional. 
+        We will create a Constant Value in the graph if User didn't provide one.
+        """
+        input_type_str1, input_shape_str1 = (
+            self._get_input_type_and_shape_strs(graph, 1)
+            if self.input_names[1] in graph_input_names
+            else ("int64", "0")
+        )
+        input_type_str2, input_shape_str2 = (
+            self._get_input_type_and_shape_strs(graph, 2)
+            if self.input_names[2] in graph_input_names
+            else ("float", "0")
+        )
+        input_type_str3, input_shape_str3 = (
+            self._get_input_type_and_shape_strs(graph, 3) if self.input_names[3] in graph_input_names else ("bool", "0")
+        )
+        input_type_str4, input_shape_str4 = (
+            self._get_input_type_and_shape_strs(graph, 4) if self.input_names[4] in graph_input_names else ("bool", "0")
+        )
+        input_type_str5, input_shape_str5 = (
+            self._get_input_type_and_shape_strs(graph, 5) if self.input_names[5] in graph_input_names else ("bool", "0")
+        )
+
+        output_shape_str = "batch_size,max_seq_len"
```

Comment:
I don't think it's safe to hardcode arbitrary names for dimensions as they're essentially variable names that must have the same value throughout the entire graph. to do that you'd have to guarantee that no other place used the same name for a different value.

always use the dimension name from the input that uses the dynamic dimension where possible. if the dynamic dimension has no matching input value either use an empty string so the rank is correct but it's an unnamed dynamic dimension, or figure out a safe way to come up with a unique name (potentially based on input dim names). 

also in this case 'batch_size' seems inappropriate as the input to the step isn't batched and it's more the number of sentences isn't it so 'num_sentences' is possibly a better name (although I'd probably append  f'_step_{step_num}' to it to make it far more unique.

Reply:
Let me add a prefix, fetching the current timestamp.